### PR TITLE
Separate pod priority from preemption

### LINKED ISSUE #62068: Priority Scheduling w/o Preemption
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind feature

**What happened**:
We are looking to use our Kubernetes cluster as a workload processing system for our clients. We want to use the priority functionality in 1.9. However, we have a hard requirement where we cannot preempt a clients job for another job. Is there any way to disable preemption but maintain priority? We only need priority to order the scheduling of workloads going onto the cluster. 

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

@jennac

### PR DESCRIPTION
**Pull request by resouer**
**What this PR does / why we need it**:
Users request to split priority and preemption feature gate so they can use priority separately.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #62068 

**Special notes for your reviewer**:

~~I kept use `ENABLE_POD_PRIORITY` as ENV name for gce cluster scripts for backward compatibility reason. Please let me know if other approach is preffered.~~

~~This is a potential **break change** as existing clusters will be affected, we may need to include this in 1.11 maybe?~~

TODO: update this doc https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/

[Update] Usage: in config file for scheduler:
```yaml
apiVersion: componentconfig/v1alpha1
kind: KubeSchedulerConfiguration
...
disablePreemption: true
```

**Release note**:

```release-note
Split PodPriority and PodPreemption feature gate
```

### CODE REVIEW

**Comment by CaoShuFeng** [comment]
> This is a potential break change as existing clusters will be affected, we may need to include this in 1.11 maybe?

If we do not introduce a new featrue gate, but add a new argument to kube-controller-manager, to disable preemption?
like
```
--disable-pod-preemption
```

File: pkg/features/kube_features.go [hunk 1]
```diff
@@ -116,6 +116,12 @@ const (
 	// Add priority to pods. Priority affects scheduling and preemption of pods.
 	PodPriority utilfeature.Feature = "PodPriority"
 
+	// owner: @bsalamat
+	// alpha: v1.8
+	//
+	// Allow preemption of pods with lower priority. Only work with PodPriority is enabled.
+	PodPreemption utilfeature.Feature = "PodPreemption"
```
Comment (bsalamat) [inline pkg/features/kube_features.go]: Enabling preemption without enabling priority is wrong configuration. I think we should check and throw error if it happens. I don't know if we have any standard location in our codebase to check for such dependencies of feature gates.

**Comment by bsalamat** [comment]
> What about we do not introduce a new featrue gate, but add a new argument to kube-controller-manager, to disable preemption?

I like this idea. We can then consider preemption enabled only when pod priority is enabled and preemption is not explicitly disabled. It prevents bad configurations where priority is disabled and preemption is enabled.

**Comment by resouer** [comment]
> What about we do not introduce a new feature gate, but add a new argument to kube-controller-manager, to disable preemption?

Sorry, but why `kube-controller-manager`, shouldn't this be part of scheduler configure?

**Comment by CaoShuFeng** [comment]
> Sorry, but ... why kube-controller-manager? shouldn't this be part of scheduler configure?

kube-scheduler, I wrote it  incorrectly. :smile:

File: pkg/features/kube_features.go (hunk 1 above)
Comment (resouer) [reply pkg/features/kube_features.go]: Thanks, I will make it a scheduler configuration.

**Comment by bsalamat** [comment]
Thanks, Harry. Looks generally good, but could you please add a test for the case that preemption is disabled?

**Comment by resouer** [comment]
@bsalamat Thanks for review, will do.

File: test/integration/scheduler/preemption_test.go [hunk 2]
```diff
@@ -285,6 +285,92 @@ func TestPreemption(t *testing.T) {
...
+			}
+		}
+		// Create the "pod".
+		preemptor, err := createPausePod(cs, test.pod)
+		if err != nil {
+			t.Errorf("Error while creating high priority pod: %v", err)
+		}
+		// Ensure no pod is preempted.
+		for _, p := range pods {
+			if err = wait.Poll(time.Second, wait.ForeverTestTimeout, podIsGettingEvicted(cs, p.Namespace, p.Name)); err == nil {
```
Comment (bsalamat) [inline test/integration/scheduler/preemption_test.go]: This causes the test to take a long time (30 seconds for each pod) before succeeding and has the potential of timing out when there are more pods in the test. Please use `waitForPodUnschedulable` to check that preemptor is marked unschedulable and then you can use `waitForNominatedNodeName` on the preemptor to ensure that its `NominatedNodeName` is never set. I'd suggest creating a variation of `waitForNominatedNodeName` that takes a timeout. Then you can use the new variation and wait only several seconds to ensure that the nominated node name is not set.

Comment (resouer) [reply test/integration/scheduler/preemption_test.go]: Thanks for pointing out. Just fixed.

File: test/integration/scheduler/util.go [hunk 3]
```diff
@@ -63,7 +63,8 @@ type TestContext struct {
 	scheduler              *scheduler.Scheduler
 }
 
-// createConfiguratorWithPodInformer create a configurator for scheduler with given informer factory, custom name and  pod informer.
+// createConfiguratorWithPodInformer create a configurator for scheduler with given informer factory,
```
Comment (bsalamat) [inline test/integration/scheduler/util.go]: s/create/creates/
Please remove "with given informer factory, custom name and pod informer" from the comment. They don't add any value to the comment and need to be updated when function args change.

**Review by bsalamat — COMMENTED** [review]
Thanks, @resouer! Looks good. Just a minor comment.

**Review by bsalamat — COMMENTED** [review]
/lgtm

Thanks, @resouer!

**Comment by resouer** [comment]
/assign @timothysc 

for test approval

**Comment by resouer** [comment]
/assign @mikedanese

For component config approval

**Pull request merged**